HW0/P4/P4.py

```diff
@@ -30,47 +30,92 @@
     # Load true trajectory and plot it
     # Normally, this data wouldn't be available in the real world
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '--b', label='True trajectory')
+    s_true = np.loadtxt("P4_trajectory.txt", delimiter=',')
+    x_coords = s_true[:,0]
+    y_coords = s_true[:,1]
+    z_coords = s_true[:,2]
+    ax.plot(x_coords, y_coords, z_coords, '--b', label='True trajectory')
 
     #####################
     # Part 2:
     #
     # Read the observation array and plot it (Part 2)
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    m_k = np.loadtxt("P4_measurements.txt", delimiter=',')
+    r_matrix = np.array([[1/rx, 0, 0], [0, 1/ry, 0], [0, 0, 1/rz]])
+    x_k = np.dot(m_k, r_matrix)
+    x_coords = x_k[:,0]
+    y_coords = x_k[:,1]
+    z_coords = x_k[:,2]
+    ax.plot(x_coords, y_coords, z_coords, '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
-
-    # A = ?
-    # a = ?
-    # s = ?
+    A = np.array([[1,0,0,dt,0,0], [0,1,0,0,dt,0], [0,0,1,0,0,dt], [0,0,0,1-c*dt,0,0], [0,0,0,0,1-c*dt,0], [0,0,0,0,0,1-c*dt]])
+    a = np.transpose(np.matrix([0,0,0,0,0,g*dt]))
+    s = np.asmatrix(np.zeros([6,K]))
 
     # Initial conditions for s0
     # Compute the rest of sk using Eq (1)
+    s_0 = np.transpose(np.matrix([0,0,2,15,3.5,4.0]))
+    s[:,0] = s_0
+    for i in xrange(1,K):
+        # Update the ith column in s according the (i-1)th column
+        s[:,i] = A * s[:,i-1] + a
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    # Convert s back to array; take first 3 rows as x,y,z coords respectively
+    s = np.asarray(s)
+    x_coords = s[0,:]
+    y_coords = s[1,:]
+    z_coords = s[2,:]
+
+    ax.plot(x_coords, y_coords, z_coords, '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    A = np.asmatrix(A)
+    B = np.asmatrix(np.diag([bx,by,bz,bvx,bvy,bvz]))
+    C = np.matrix([[rx,0,0,0,0,0], [0,ry,0,0,0,0], [0,0,rz,0,0,0]])
 
     # Initial conditions for s0 and Sigma0
-    # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    s_0 = np.transpose(np.matrix([0,0,2,15,3.5,4.0]))
+    Sigma0 = 0.01 * np.identity(6)
+    s = np.asmatrix(np.zeros([6,K]))
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    def predictS(s_k):
+        return np.asmatrix(A) * s_k + a
```

Comment:
being nitpicky but you set A as a matrix earlier in line 81 so think this is redundant? There's nothing else I can see to comment on lol